internal/daemon/controller/handlers/authmethods/ldap.go

```diff
@@ -0,0 +1,364 @@
+package authmethods
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/hashicorp/boundary/internal/auth/ldap"
+	ldapstore "github.com/hashicorp/boundary/internal/auth/ldap/store"
+	"github.com/hashicorp/boundary/internal/daemon/controller/auth"
+	"github.com/hashicorp/boundary/internal/daemon/controller/handlers"
+	"github.com/hashicorp/boundary/internal/errors"
+	pbs "github.com/hashicorp/boundary/internal/gen/controller/api/services"
+	"github.com/hashicorp/boundary/internal/types/action"
+	pb "github.com/hashicorp/boundary/sdk/pbs/controller/api/resources/authmethods"
+	"google.golang.org/grpc/codes"
+)
+
+var ldapMaskManager handlers.MaskManager
+
+func init() {
+	var err error
+	if ldapMaskManager, err = handlers.NewMaskManager(handlers.MaskDestination{&ldapstore.AuthMethod{}}, handlers.MaskSource{&pb.AuthMethod{}, &pb.LdapAuthMethodAttributes{}}); err != nil {
+		panic(err)
+	}
+
+	IdActions[ldap.Subtype] = action.ActionSet{
+		action.NoOp,
+		action.Read,
+		action.Update,
+		action.Delete,
+		action.ChangeState,
+		action.Authenticate,
+	}
+}
+
+const (
+	urlsField                 = "attributes.urls"
+	bindDnField               = "attributes.bind_dn"
+	bindPasswordField         = "attributes.bind_password"
+	clientCertificateField    = "attributes.client_certificate"
+	clientCertificateKeyField = "attributes.client_certificate_key"
+	certificatesField         = "attributes.certificates"
+	accountAttributesMapField = "attributes.account_attribute_maps"
+)
+
+func (s Service) authenticateLdap(ctx context.Context, req *pbs.AuthenticateRequest, authResults *auth.VerifyResults) (*pbs.AuthenticateResponse, error) {
+	const op = "authmethod_service.(Service).authenticateLdap"
+	if req == nil {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "Nil request.")
+	}
+	if authResults == nil {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "Nil auth results.")
+	}
+	reqAttrs := req.GetLdapLoginAttributes()
+
+	ldapRepo, err := s.ldapRepoFn()
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	ldapFn := func() (ldap.Authenticator, error) {
+		return ldapRepo, nil
+	}
+
+	iamRepo, err := s.iamRepoFn()
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	iamFn := func() (ldap.LookupUser, error) {
+		return iamRepo, nil
+	}
+
+	atRepo, err := s.atRepoFn()
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	atFn := func() (ldap.AuthTokenCreator, error) {
+		return atRepo, nil
+	}
+
+	rawTk, err := ldap.Authenticate(ctx, ldapFn, iamFn, atFn, req.GetAuthMethodId(), reqAttrs.GetLoginName(), reqAttrs.GetPassword())
+	if err != nil {
+		// let's not send back too much info about the error
```

Comment:
Is the full error captured in the audit logs or does this truncate that error too?

Reply:
audit events were/are generated when the error is originally created so it's fully captured.